subxt/src/metadata/metadata_type.rs

```diff
@@ -68,28 +70,48 @@ pub enum MetadataError {
     #[error("Failed to decode constant value: {0}")]
     ConstantValueError(CodecError),
     /// Constant is not in metadata.
-    #[error("Constant {0} not found")]
-    ConstantNotFound(&'static str),
+    #[error("Constant not found")]
+    ConstantNotFound,
     /// Type is not in metadata.
     #[error("Type {0} missing from type registry")]
     TypeNotFound(u32),
+    /// Runtime pallet metadata is incompatible with the static one.
+    #[error("Pallet {0} has incompatible metadata")]
+    IncompatiblePalletMetadata(&'static str),
+    /// Runtime metadata is not fully compatible with the static one.
+    #[error("Node metadata is not fully compatible")]
+    IncompatibleMetadata,
 }
 
 /// Runtime metadata.
 #[derive(Clone, Debug)]
 pub struct Metadata {
-    metadata: RuntimeMetadataLastVersion,
+    inner: Arc<MetadataInner>,
+}
+
+// We hide the innards behind an Arc so that it's easy to clone and share.
+#[derive(Debug)]
+struct MetadataInner {
+    metadata: RuntimeMetadataV14,
     pallets: HashMap<String, PalletMetadata>,
     events: HashMap<(u8, u8), EventMetadata>,
     errors: HashMap<(u8, u8), ErrorMetadata>,
+    // The hashes uniquely identify parts of the metadata; different
+    // hashes mean some type difference exists between static and runtime
+    // versions. We cache them here to avoid recalculating:
+    cached_metadata_hash: RwLock<Option<[u8; 32]>>,
+    cached_call_hashes: HashCache,
+    cached_constant_hashes: HashCache,
+    cached_storage_hashes: HashCache,
 }
 
 impl Metadata {
     /// Returns a reference to [`PalletMetadata`].
     pub fn pallet(&self, name: &'static str) -> Result<&PalletMetadata, MetadataError> {
-        self.pallets
+        self.inner
+            .pallets
             .get(name)
-            .ok_or_else(|| MetadataError::PalletNotFound(name.to_string()))
```

Comment:
isn't this useful to know which pallet that was not found in the error? Or am I missing something